### PR TITLE
Refactor and fix queue resharding

### PR DESCRIPTION
**Pull request by tomwilkie**
Right now just refactoring to try and simplify the code.  Then I'll expose a lot more of the resharding logic as metrics, then try and fix. /cc @cstyan

### CODE REVIEW

File: storage/remote/queue_manager.go [hunk 1]
```diff
@@ -731,17 +678,14 @@ func (s *shards) sendSamplesWithBackoff(ctx context.Context, samples []prompb.Ti
 
 		if err == nil {
 			succeededSamplesTotal.WithLabelValues(s.qm.queueName).Add(float64(len(samples)))
-			now := time.Now()
-			s.qm.setLastSendTimestamp(now)
-			s.qm.setHighestSentTimestamp(highest)
+			s.qm.highestSentTimestampMetric.Set(float64(highest / 1000))
```
Comment (cstyan) [inline storage/remote/queue_manager.go]: are you sure we don't want to check that highest > the value we last set for `highestSentTimestampMetric` first?

Comment (cstyan) [reply storage/remote/queue_manager.go]: ah I see, `maxGauge` :+1:

File: storage/remote/queue_manager.go [hunk 2]
```diff
@@ -425,7 +389,6 @@ func (t *QueueManager) updateShardsLoop() {
 }
 
 func (t *QueueManager) calculateDesiredShards() {
-	t.samplesIn.tick()
```
Comment (gouthamve) [inline storage/remote/queue_manager.go]: Is this the fix for the issue?

File: storage/remote/queue_manager.go [hunk 3]
```diff
@@ -731,17 +678,14 @@ func (s *shards) sendSamplesWithBackoff(ctx context.Context, samples []prompb.Ti
...
-			s.qm.setHighestSentTimestamp(highest)
+			s.qm.highestSentTimestampMetric.Set(float64(highest / 1000))
 			return nil
 		}
 
 		if _, ok := err.(recoverableError); !ok {
 			return err
 		}
 		retriedSamplesTotal.WithLabelValues(s.qm.queueName).Add(float64(len(samples)))
-		level.Error(s.qm.logger).Log("err", err)
```
[the remark this answers is not in the document]
Comment (tomwilkie) [reply storage/remote/queue_manager.go]: Yeah, some retryable errors won't be logged.  I thought it was a bit spammy.

File: storage/remote/queue_manager.go (hunk 2 above)
Comment (tomwilkie) [reply storage/remote/queue_manager.go]: This is a correctness fix; we shouldn't be ticking this N times, one for each queue.  This isn't the main fix though.

**Comment by tomwilkie** [comment]
> Changes look sane but what was the issue and how was it fixed by this? As far as I can see this is just moving and adding some metrics around.

Initial changes were just moving stuff around, sorry.  Made it easier for me to see the real bug.

The real issue was that using two rates (one measure before appending to the WAL, one after) was never going to work - if you get behind, the reading from the WAL rate will go high and the writing to rate will stay the same, and indicate you need to scale down - which is not what you want to do when you're behind!

What we do now is use the two timestamps we track - the max appending to the WAL and the max sent.  We multiply the difference by the rate at which we are appending to WAL to estimate the number of pending samples. 

Another fix was to measure the ratio of dropped samples and not the number of dropped samples, and then use this in calculating the pending samples.  This fixes queue that make heavy use of relabelling.

**Comment by tomwilkie** [comment]
screenshot 2019-03-01 at 08 55 43

File: storage/remote/queue_manager.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (gouthamve) [reply storage/remote/queue_manager.go]: Can we make it a debug log? This is going to be useful when debugging upstream issues.

File: storage/remote/queue_manager.go [hunk 4]
```diff
@@ -437,9 +398,11 @@ func (t *QueueManager) calculateDesiredShards() {
 	var (
 		samplesIn          = t.samplesIn.rate()
 		samplesOut         = t.samplesOut.rate()
-		samplesDropped     = t.samplesDropped.rate()
-		samplesPending     = samplesIn - samplesDropped - samplesOut
+		samplesKeptRatio   = samplesOut / (t.samplesDropped.rate() + samplesOut)
 		samplesOutDuration = t.samplesOutDuration.rate()
+		highestSent        = t.highestSentTimestampMetric.value
```
Comment (gouthamve) [inline storage/remote/queue_manager.go]: This is protected by a mutex, need to use that. Maybe add a `Get` method to `maxGauge`

**Comment by gouthamve** [comment]
LGTM apart from the two comments.

File: storage/remote/queue_manager.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomwilkie) [reply storage/remote/queue_manager.go]: Done

File: storage/remote/queue_manager.go [hunk 5]
```diff
@@ -437,9 +398,11 @@ func (t *QueueManager) calculateDesiredShards() {
...
-		samplesDropped     = t.samplesDropped.rate()
-		samplesPending     = samplesIn - samplesDropped - samplesOut
+		samplesKeptRatio   = samplesOut / (t.samplesDropped.rate() + samplesOut)
 		samplesOutDuration = t.samplesOutDuration.rate()
+		highestSent        = t.highestSentTimestampMetric.Get()
+		highestRecv        = highestTimestamp.Get()
+		samplesPending     = (highestRecv - highestSent) * samplesIn * samplesKeptRatio
 	)
 
 	// We use an integral accumulator, like in a PID, to help dampen oscillation.
```
Comment (cstyan) [inline storage/remote/queue_manager.go]: I think we can remove the integral accumulator now

**Pull request merged**